packages/core/src/node/process/process.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from "inversify";
+import { ILogger } from '../../common/logger';
+import * as child from 'child_process';
+import * as stream from 'stream';
+import { Emitter, Event } from '../../common/event';
+
+export interface IProcessExitEvent {
+    code: number,
+    signal?: string
+}
+
+@injectable()
+export abstract class Process {
+
+    abstract readonly type: 'Raw' | 'Terminal';
+    abstract pid: number;
+    abstract output: stream.Readable;
+    protected abstract process: child.ChildProcess | undefined;
+    protected abstract terminal: any;
+    protected exitEmitter = new Emitter<IProcessExitEvent>();
```

Comment:
`readonly`?

I wonder `Process` is not `Disposable`?  E.g. emitters should be disposed when a process is disposed.

Reply:
Right readonly.

About the Disposable, I'm glad we get to talk about this since I'm not clear about the use case for it and it's been a long standing question I have...


I tought in typescript/javascript you would get automatic refcounting and garbage collection no ?

So as all the references to a process object would be gone it would get destructed along with all it's properties... ?

I thought the disposable was more for like if the object were to hold an OS resource like a socket that you would want to be cleaned up right away.

Could you please explain a bit more the Disposables ?

Thanks!

Reply:
I think your understanding is correct: https://code.visualstudio.com/docs/extensionAPI/patterns-and-principles#_disposables.

And in this context, it is not necessary.
 and in this case `Disposable` is not necessary.

Reply:
OK thanks for the link!

So I should however dispose my event listeners in the tests.

I'll fix that

Reply:
Thinking about this more actually since I register an event handler on the same class the emitter is in there's nothing to cleanup that listener so the Process class has a circular reference to itself.

So I need to make Process disposable and unregister all listeners including itself by calling each emitter dispose method.

Reply:
I thought js use mark-and-sweep for GC and able to clean up dangling object graphs, but you better double check it :)